app/controllers/reports_controller.rb

```diff
@@ -35,38 +35,16 @@ def edit
   # POST /reports.json
   def create
     @report = Report.new(report_params)
-    @report.pipeline_output.taxon_counts.each do |taxon_count|
-      a = {}
-      a[:tax_id] = taxon_count.tax_id
-      a[:tax_level] = taxon_count.tax_level
-      a[:name] = taxon_count.name
-      a[:rpm] = 1e6 * taxon_count.count.to_f / @report.pipeline_output.total_reads
-      a[:hit_type] = taxon_count.count_type
-      normalized_count = taxon_count.count.to_f / @report.pipeline_output.total_reads
-      sum = 0
-      sum_sq = 0
-      n = 0
-      @report.background.pipeline_outputs.each do |bg_pipeline_output|
-        bg_taxon_count = bg_pipeline_output.taxon_counts.find_by(tax_id: taxon_count.tax_id, count_type: taxon_count.count_type)
-        if bg_taxon_count
-          bg_count = bg_taxon_count.count
-          normalized_bg_count = bg_count.to_f / bg_pipeline_output.total_reads
-        else
-          normalized_bg_count = 0
-        end
-        sum += normalized_bg_count
-        sum_sq += normalized_bg_count**2
-        n += 1
-      end
-      mean = sum.to_f / n
-      stdev = Math.sqrt((sum_sq.to_f - sum**2 / n) / (n - 1))
-      a[:zscore] = if stdev > 0
-                     (normalized_count - mean) / stdev
-                   else
-                     0
-                   end
-      @report.taxon_zscores.new(a)
+    unless Background.find_by(name: "default") || @report.background
+      raise "No background specified and background 'default' does not exist"
     end
+    @report.background = Background.find_by(name: "default") unless @report.background
+    total_reads = @report.pipeline_output.total_reads
+    summary = @report.background.summarize
+    data = @report.pipeline_output.taxon_counts.map { |h| h.attributes.merge(norm_count: h["count"] / total_reads.to_f) }
+    data_and_background = (data + summary).group_by { |h| [h["tax_id"], h["tax_level"], h["name"], h["count_type"]] }.map { |_k, v| v.reduce(:merge) }
```

Comment:
wouldn't you get a bunch of records without norm_count if there's no count for the pipeline output?  In that case, I believe you don't need to generate the z_score for those.